plugins/inputs/ctrlx_datalayer_sse/ctrlx_datalayer_sse.go

```diff
@@ -0,0 +1,498 @@
+package ctrlx_datalayer_sse
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal/choice"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+// This plugin is based on the official ctrlX CORE API. Documentation can be found in OpenAPI format at:
+// https://boschrexroth.github.io/rest-api-description/ctrlx-automation/ctrlx-core/
+// Used APIs are:
+// * ctrlX CORE - Authorization and Authentication API
+// * ctrlX CORE - Data Layer API
+//
+// All communication between the device and this input plugin is based
+// on https REST and HTML5 Server Sent Events (sse).
+
+const (
+	defaultKeepaliveInterval          = config.Duration(60 * time.Second)
+	defaultErrorInterval              = config.Duration(10 * time.Second)
+	defaultReconnectInterval          = config.Duration(10 * time.Second)
+	defaultCreateSubscriptionInterval = config.Duration(2 * time.Second)
+	defaultPublishInterval            = config.Duration(1 * time.Second)
+	defaultSamplingInterval           = config.Duration(1 * time.Second)
+	defaultTimeout                    = config.Duration(5 * time.Second)
+	defaultQueueSize                  = 10
+	defaultQueueBehaviour             = "DiscardOldest"
+	defaultValueChange                = "StatusValue"
+	defaultMeasurementName            = "ctrlx"
+)
+
+var queueBehaviours = []string{"", "DiscardOldest", "DiscardNewest"}
+var valueChanges = []string{"", "Status", "StatusValue", "StatusValueTimestamp"}
+
+type ctrlXDataLayerSSE struct {
+	acc                telegraf.Accumulator
+	connection         *http.Client
+	token              token
+	autoReconnect      bool
+	Hostname           string          `toml:"hostname"`
+	Username           string          `toml:"username"`
+	Password           string          `toml:"password"`
+	InsecureSkipVerify bool            `toml:"insecure_skip_verify"`
+	Timeout            config.Duration `toml:"timeout"`
+	Subscription       []subscription
+	Log                telegraf.Logger `toml:"-"`
+	wg                 sync.WaitGroup
+	mux                sync.Mutex
+}
+
+type subscription struct {
```

Comment:
We thought it would be sufficient to export the base struct `CtrlXDataLayerSSE` where the subscription is embedded in.

Reply:
It might be sufficient, however, we usually export the user-option structs to be able to e.g. test parsing in the `config` package or being able to fixup options later. The same holds for deprecation handling and a few other features of Telegraf. So please export this structure.